### PR TITLE
Added support for manual snapshot.

### PR DESCRIPTION
**Pull request by steve-bate**
Also added support for camera log retrieval and PTZ preset list retrieval. The log retrieval could be useful for intrusion detection. I've also added support for turning the IR LEDs on and off, for setting the IR auto/manual mode and for retrieving product information about the camera.

To support the snapshot retrieval I needed to make some small changes to allow the raw HTTP request data to be returned.

Thanks for writing this library. It's been very useful to me.

### CODE REVIEW

File: foscam.py [hunk 1]
```diff
@@ -65,6 +65,10 @@ def send_command(self, cmd, params=None):
         try:
             raw_string = ''
             raw_string = urllib.urlopen(cmdurl).read()
+            if raw:
+                if self.verbose:
+                    print 'Returning raw Foscam response: len=%d' % len(raw_string)
+                return raw_string
```
Comment (taiyangc) [inline foscam.py]: Probably better if we return the same tuple format here for consistency `(code, data)`. Other than that, this looks great! Thanks.

File: camtest.py [hunk 2]
```diff
@@ -302,5 +305,19 @@ def print_res(*args, **kwargs):
             timeout -= 0.5
         self.assertTrue(flag)
 
+
+    # *************** SnapPicture Function *******************
+
+    def test_snap_picture_2(self):
+        data = self.foscam.snap_picture_2()
+        with open("test.jpg", "wb") as fp:
```
Comment (taiyangc) [inline camtest.py]: And sorry to nitpick, please use ' for quotes throughout.

File: foscam.py [hunk 3]
```diff
@@ -65,6 +65,10 @@ def send_command(self, cmd, params=None):
         try:
             raw_string = ''
             raw_string = urllib.urlopen(cmdurl).read()
+            if raw:
+                if self.verbose:
+                    print 'Returning raw Foscam response: len=%d' % len(raw_string)
+                return FOSCAM_SUCCESS, raw_string
```
Comment (steve-bate) [inline foscam.py:71]: Added response code for raw response.

File: camtest.py [hunk 4]
```diff
@@ -313,5 +316,20 @@ def print_res(*args, **kwargs):
             timeout -= 0.5
         self.assertTrue(flag)
 
+
+    # *************** SnapPicture Function *******************
+
+    def test_snap_picture_2(self):
+        rc, data = self.foscam.snap_picture_2()
+        self.assertEqual(rc, FOSCAM_SUCCESS)
+        with open('test.jpg', 'wb') as fp:
```
Comment (steve-bate) [inline camtest.py:351]: Changed to single quotes.

File: foscam.py [hunk 5]
```diff
@@ -83,11 +87,13 @@ def send_command(self, cmd, params=None):
...
-    def execute_command(self, cmd, params=None, callback=None):
+    def execute_command(self, cmd, params=None, callback=None, raw=False):
         '''
         Execute a command and return a parsed response.
         '''
-        def execute_with_callbacks(cmd, params=None, callback=None):
+        def execute_with_callbacks(cmd, params=None, callback=None, raw=False):
+            if raw:
+                return self.send_command(cmd, params, raw)
             code, params = self.send_command(cmd, params)
```
Comment (taiyangc) [inline foscam.py]: Probably need to merge these 3 lines so callback works for raw=True.

**Comment by steve-bate** [comment]
Quotes changed. Thanks.

**Comment by taiyangc** [comment]
Thank you.

**Pull request merged**